aws/src/main/java/org/apache/iceberg/aws/s3/VendedCredentialsProvider.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.iceberg.rest.ErrorHandlers;
+import org.apache.iceberg.rest.HTTPClient;
+import org.apache.iceberg.rest.RESTClient;
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+import org.apache.iceberg.rest.auth.OAuth2Util;
+import org.apache.iceberg.rest.credentials.Credential;
+import org.apache.iceberg.rest.responses.LoadCredentialsResponse;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.cache.CachedSupplier;
+import software.amazon.awssdk.utils.cache.RefreshResult;
+
+public class VendedCredentialsProvider implements AwsCredentialsProvider, SdkAutoCloseable {
+  public static final String URI = "credentials.uri";
+  private volatile HTTPClient client;
+  private final Map<String, String> properties;
+  private final CachedSupplier<AwsCredentials> credentialCache;
+
+  private VendedCredentialsProvider(Map<String, String> properties) {
+    Preconditions.checkArgument(null != properties.get(URI), "Invalid URI: null");
+    this.properties = properties;
+    this.credentialCache =
+        CachedSupplier.builder(this::refreshCredential)
+            .cachedValueName(VendedCredentialsProvider.class.getName())
+            .build();
+  }
+
+  @Override
+  public AwsCredentials resolveCredentials() {
+    return credentialCache.get();
+  }
+
+  @Override
+  public void close() {
+    IoUtils.closeQuietly(client, null);
+    credentialCache.close();
+  }
+
+  public static VendedCredentialsProvider create(Map<String, String> properties) {
+    return new VendedCredentialsProvider(properties);
+  }
+
+  private RESTClient httpClient() {
+    if (null == client) {
+      synchronized (this) {
+        if (null == client) {
+          client = HTTPClient.builder(properties).uri(properties.get(URI)).build();
+        }
+      }
+    }
+
+    return client;
+  }
+
+  private LoadCredentialsResponse fetchCredentials() {
+    return httpClient()
+        .get(
+            properties.get(URI),
+            null,
+            LoadCredentialsResponse.class,
+            OAuth2Util.authHeaders(properties.get(OAuth2Properties.TOKEN)),
+            ErrorHandlers.defaultErrorHandler());
+  }
+
+  private RefreshResult<AwsCredentials> refreshCredential() {
+    LoadCredentialsResponse response = fetchCredentials();
+
+    List<Credential> s3Credentials =
+        response.credentials().stream()
+            .filter(c -> c.prefix().startsWith("s3"))
+            .collect(Collectors.toList());
+
+    Preconditions.checkState(!s3Credentials.isEmpty(), "Invalid S3 Credentials: empty");
+
+    Optional<Credential> credentialWithPrefix =
+        s3Credentials.stream().max(Comparator.comparingInt(c -> c.prefix().length()));
```

Comment:
[doubt] how will this work for credentials which are expanding different prefixes :
for ex : 
s3//abc/123
s3//xyz/1

and the request is for s3://abc ? 

Never the less how will this work for cases with same prefixes : 
1. s3://abc/prefix-1/
2. s3://abc/prefix-123/ 
we will use prefix-123 here instead of prefix-1 ? for the calls length is not the correct way to find the credentials imho

please let me know your thoughts.

Reply:
This is actually a good callout and I think an oversight form what we're trying to define in the spec.  Here we are taking the most selective prefix, but that doesn't necessarily correspond to the prefix we're using the credential against.

Reply:
In order to get the behavior we're expecting we would probably need to update `S3FileIO::client` to take the path as an argument so that the client can be configured/returned with the correct credential.  Thoughts @nastra?  I assume there  is a similar issue on the GCP side.

Reply:
@singhpk234 to answer your specific question, it should fail because there is no prefix in your example that covers the requested path.

Reply:
@danielcweeks  got you, This is always been a challenge when using credential provider as if you see it's `resolve credenitial()` doesn't take any args so to wire in correct creds for correct paths.

Reply:
Yeah the credential to use is the credential with the longest matching prefix for a _given requested path_.  In the case that there's no credential with a matching prefix for a given path, I think we'd want to just fail on the client side; s3 or whatever storage system would fail the request anyways but I think we may as well avoid the request if we know that no credential could cover the requested path.

Reply:
I think what we rather might want to do going forward is only allow the server to return a single credential per provider. That way it's easier for the client

Reply:
Hm, sure that's easier for the client but I'm not sure about making the protocol unnecessarily restrictive. Though I do understand in most cases a single credential per provider is enough, I just want to make sure we're not putting the spec in an awkward spot in case there are legitimate use cases of different credentials per prefix. In general, we've always followed a pattern of avoiding putting unnecessary restrictions on protocol. Also I don't know if it's really that much more complex for clients to resolve the longest matching prefix for a given path

Reply:
I agree that it's reasonable to enforce a single prefix for provider since that was the initial ask.  There's an opportunity to improve this, but we can follow up with added support.  I don't think it's too difficult, but there are a few edge cases that we'd need to work through.  I'm ok proceeding with this implementation.

Reply:
I think I'm good as long as this is just constrained to the current implementation and we're not requiring any server side behavior in the spec